cpp/CMakeLists.txt

```diff
@@ -784,27 +632,103 @@ endif(HT_DEFAULT_ALLOCATOR)
+set(CUDF_MODULES cudf_agg
+                 cudf_binop
+                 cudf_bitmask
+                 cudf_column
+                 cudf_comms
+                 cudf_copying
+                 cudf_datetime
+                 cudf_dict
+                 cudf_filling
+                 cudf_groupby
+                 cudf_hash
+                 cudf_interop
+                 cudf_io
+                 cudf_jit
+                 cudf_join
+                 cudf_lists
+                 cudf_merge
+                 cudf_part
+                 cudf_quant
+                 cudf_reduce
+                 cudf_replace
+                 cudf_rolling
+                 cudf_scalar
+                 cudf_search
+                 cudf_sort
+                 cudf_streamc
+                 cudf_strings
+                 cudf_structs
+                 cudf_table
+                 cudf_text
+                 cudf_xform
+                 cudf_xpose
+                 cudf_unop
+)
+
+foreach(MODULE IN LISTS CUDF_MODULES)
+  list(APPEND CUDF_MODULE_TARGETS "\$<TARGET_OBJECTS:${MODULE}>")
+endforeach()
+
+set(NUM_LIBRARY_PARTITIONS 1 CACHE INTERNAL "")
+message(STATUS "Number of libraries to be created: ${NUM_LIBRARY_PARTITIONS}")
+
+list(LENGTH CUDF_MODULES NUM_MODULES)
+math(EXPR VALUE ${NUM_MODULES}/${NUM_LIBRARY_PARTITIONS})
+set(NUM_MODULES_PER_LIB ${VALUE} CACHE INTERNAL "")
+math(EXPR VALUE ${NUM_MODULES}%${NUM_LIBRARY_PARTITIONS})
+set(NUM_MODULES_PER_LIB_REM ${VALUE})
+message(STATUS "Number of modules per library partition: ${NUM_MODULES_PER_LIB} with remainder ${NUM_MODULES_PER_LIB_REM}")
+
+if(${NUM_LIBRARY_PARTITIONS} GREATER_EQUAL 2)
+  math(EXPR STOP ${NUM_LIBRARY_PARTITIONS}-2)
+  foreach(INDEX RANGE ${STOP})
+    math(EXPR START_INDEX ${INDEX}*${NUM_MODULES_PER_LIB})
+    list(SUBLIST CUDF_MODULE_TARGETS ${START_INDEX} ${NUM_MODULES_PER_LIB} LIB_TARGETS)
+    message(STATUS "Targets for cudf_partition_${INDEX}: ${LIB_TARGETS}")
+    add_library("cudf_partition_${INDEX}" ${LIB_TARGETS})
+    list(APPEND CUDF_LIBRARIES "cudf_partition_${INDEX}")
+  endforeach()
+
+  math(EXPR INDEX ${NUM_LIBRARY_PARTITIONS}-1)
+  math(EXPR START_INDEX ${INDEX}*${NUM_MODULES_PER_LIB})
+  math(EXPR NMODULES ${NUM_MODULES_PER_LIB}+${NUM_MODULES_PER_LIB_REM})
+  list(SUBLIST CUDF_MODULE_TARGETS ${START_INDEX} ${NMODULES} LIB_TARGETS)
+  message(STATUS "Targets for cudf_partition_${INDEX}: ${LIB_TARGETS}")
+  add_library("cudf_partition_${INDEX}" ${LIB_TARGETS})
+  list(APPEND CUDF_LIBRARIES "cudf_partition_${INDEX}")
+else ()
+  message(STATUS "Targets for libcudf: ${CUDF_MODULE_TARGETS}")
+  add_library("cudf" ${CUDF_MODULE_TARGETS})
+  list(APPEND CUDF_LIBRARIES "cudf")
+endif()
+
+message(STATUS "cuDF library partitions: ${CUDF_LIBRARIES}")
 
 # link targets for cuDF
-target_link_libraries(cudf rmm arrow arrow_cuda nvrtc ${CUDART_LIBRARY} cuda ${ZLIB_LIBRARIES} ${Boost_LIBRARIES})
+foreach(CUDF_LIBRARY IN LISTS CUDF_LIBRARIES)
```

Comment:
@jrhemstad @harrism are there significant performance regressions associated with having multiple library target like this?

Reply:
Linker performance? Runtime performance? I don't know the answer.

Reply:
@mt-jones Are you asking about compile-time performance, or run-time performance?

Reply:
Run-time performance.

Reply:
Not that I know of.